File: packages/application/src/process-invoice-reminders.use-case.ts
import { Invoice } from '@monolegal/domain';
import type { IEmailProvider } from '@monolegal/domain';
import type { IInvoiceRepository } from '@monolegal/domain';
import { REMINDER_STATUSES, type ILogger } from '@monolegal/shared';

export interface ProcessInvoiceRemindersResult {
  processed: number;
  failed: number;
}

export class ProcessInvoiceRemindersUseCase {
  constructor(
    private readonly invoiceRepository: IInvoiceRepository,
    private readonly emailProvider: IEmailProvider,
    private readonly logger: ILogger,
  ) {}

  async execute(): Promise<ProcessInvoiceRemindersResult> {
    const invoices = await this.invoiceRepository.findByStatus(REMINDER_STATUSES);

    this.logger.info('Starting invoice reminder processing', {
      totalInvoices: invoices.length,
    });

    let processed = 0;
    let failed = 0;

    for (const invoiceProps of invoices) {
      try {
        await this.processInvoice(invoiceProps);
        processed++;
      } catch (error) {
        failed++;
        this.logger.error('Failed to process invoice reminder', {
          invoiceId: invoiceProps.id,
          clientName: invoiceProps.clientName,
          status: invoiceProps.status,
          error: error instanceof Error ? error.message : String(error),
        });
      }
    }

    this.logger.info('Invoice reminder processing completed', { processed, failed });

    return { processed, failed };
  }

  private async processInvoice(invoiceProps: Parameters<typeof Invoice.fromProps>[0]): Promise<void> {
    const invoice = Invoice.fromProps(invoiceProps);
    const nextStatus = invoice.getNextStatusAfterReminder();

    const emailContent = invoice.canSendFirstReminder()
      ? invoice.buildFirstReminderEmail()
      : invoice.buildSecondReminderEmail();

    await this.emailProvider.sendReminder({
      to: invoice.clientEmail,
      subject: emailContent.subject,
      body: emailContent.body,
    });

    await this.invoiceRepository.updateStatus(invoice.id, nextStatus);

    this.logger.info('Invoice reminder sent and status updated', {
      invoiceId: invoice.id,
      clientName: invoice.clientName,
      previousStatus: invoice.status,
      newStatus: nextStatus,
    });
  }
}

File: backend/dotnet/src/Monolegal.Application/ProcessInvoiceRemindersUseCase.cs
using Monolegal.Domain.Entities;
using Monolegal.Domain.Errors;
using Monolegal.Domain.Ports;
using Monolegal.Shared;

namespace Monolegal.Application;

public record ProcessInvoiceRemindersResult(int Processed, int Failed);

public class ProcessInvoiceRemindersUseCase(
    IInvoiceRepository invoiceRepository,
    IClientRepository clientRepository,
    IEmailProvider emailProvider,
    ILogger logger)
{
    public async Task<ProcessInvoiceRemindersResult> ExecuteAsync(CancellationToken cancellationToken = default)
    {
        var invoices = await invoiceRepository.FindByStatusAsync(InvoiceStatusValues.ReminderStatuses, cancellationToken);

        logger.Info("Starting invoice reminder processing", new Dictionary<string, object?>
        {
            ["totalInvoices"] = invoices.Count
        });

        var processed = 0;
        var failed = 0;

        foreach (var invoice in invoices)
        {
            try
            {
                await ProcessInvoiceAsync(invoice, cancellationToken);
                processed++;
            }
            catch (Exception ex)
            {
                failed++;
                logger.Error("Failed to process invoice reminder", new Dictionary<string, object?>
                {
                    ["invoiceId"] = invoice.Id,
                    ["clientId"] = invoice.ClientId,
                    ["status"] = InvoiceStatusValues.ToValue(invoice.Status),
                    ["error"] = ex.Message
                });
            }
        }

        logger.Info("Invoice reminder processing completed", new Dictionary<string, object?>
        {
            ["processed"] = processed,
            ["failed"] = failed
        });

        return new ProcessInvoiceRemindersResult(processed, failed);
    }

    public async Task<ProcessInvoiceRemindersResult> ExecuteForInvoiceIdAsync(string invoiceId, CancellationToken cancellationToken = default)
    {
        var invoice = await invoiceRepository.FindByIdAsync(invoiceId, cancellationToken);
        if (invoice is null)
            throw new InvoiceNotFoundError($"Invoice not found: {invoiceId}");

        if (!InvoiceStatusValues.ReminderStatuses.Contains(invoice.Status))
            throw new InvoiceTransitionError(
                $"Invoice {invoiceId} is not eligible for reminder processing (status: {InvoiceStatusValues.ToValue(invoice.Status)})");

        logger.Info("Starting single invoice reminder processing", new Dictionary<string, object?> { ["invoiceId"] = invoiceId });

        try
        {
            await ProcessInvoiceAsync(invoice, cancellationToken);
            logger.Info("Single invoice reminder processing completed", new Dictionary<string, object?>
            {
                ["invoiceId"] = invoiceId,
                ["processed"] = 1
            });
            return new ProcessInvoiceRemindersResult(1, 0);
        }
        catch (Exception ex)
        {
            logger.Error("Failed to process invoice reminder", new Dictionary<string, object?>
            {
                ["invoiceId"] = invoiceId,
                ["clientId"] = invoice.ClientId,
                ["status"] = InvoiceStatusValues.ToValue(invoice.Status),
                ["error"] = ex.Message
            });
            return new ProcessInvoiceRemindersResult(0, 1);
        }
    }

    private async Task ProcessInvoiceAsync(Invoice invoice, CancellationToken cancellationToken)
    {
        var client = await clientRepository.FindByIdAsync(invoice.ClientId, cancellationToken);
        if (client is null)
            throw new ClientNotFoundError($"Client not found for invoice {invoice.Id}: {invoice.ClientId}");

        var payload = invoice.BuildReminderPayload(client.Name);

        await emailProvider.SendReminderAsync(
            new EmailMessage(client.Email, payload.Email.Subject, payload.Email.Body),
            cancellationToken);

        try
        {
            await invoiceRepository.UpdateStatusAsync(invoice.Id, payload.NextStatus, cancellationToken);
        }
        catch (Exception ex)
        {
            logger.Warn("Email sent but status update failed", new Dictionary<string, object?>
            {
                ["invoiceId"] = invoice.Id,
                ["clientId"] = invoice.ClientId,
                ["previousStatus"] = InvoiceStatusValues.ToValue(invoice.Status),
                ["targetStatus"] = InvoiceStatusValues.ToValue(payload.NextStatus),
                ["emailAlreadySent"] = true,
                ["error"] = ex.Message
            });
            throw;
        }

        logger.Info("Invoice reminder sent and status updated", new Dictionary<string, object?>
        {
            ["invoiceId"] = invoice.Id,
            ["clientId"] = invoice.ClientId,
            ["clientEmail"] = client.Email,
            ["previousStatus"] = InvoiceStatusValues.ToValue(invoice.Status),
            ["newStatus"] = InvoiceStatusValues.ToValue(payload.NextStatus)
        });
    }
}
